drivers/storage/ide/pciidex/pdo.c

```diff
@@ -1,537 +1,851 @@
-	{
-		*BusMasterPortBase = 0;
-	}
-
-	if ((PciConfig.ProgIf >> BaseIndex) & 0x1)
-	{
-		/* Native mode */
-		if ((PciConfig.u.type0.BaseAddresses[BaseIndex + 0] & PCI_ADDRESS_IO_SPACE) &&
-		    (PciConfig.u.type0.BaseAddresses[BaseIndex + 1] & PCI_ADDRESS_IO_SPACE))
-		{
-			/* Channel is enabled */
-			*CommandPortBase = PciConfig.u.type0.BaseAddresses[BaseIndex + 0] & PCI_ADDRESS_IO_ADDRESS_MASK;
-			*ControlPortBase = PciConfig.u.type0.BaseAddresses[BaseIndex + 1] & PCI_ADDRESS_IO_ADDRESS_MASK;
-			*InterruptVector = PciConfig.u.type0.InterruptLine;
-			ret = STATUS_SUCCESS;
-		}
-	}
-	else
-	{
-		/* Compatibility mode */
-		switch (DeviceExtension->Channel)
-		{
-			case 0:
-				if (IoGetConfigurationInformation()->AtDiskPrimaryAddressClaimed)
-					ret = STATUS_INSUFFICIENT_RESOURCES;
-				else
-				{
-					*CommandPortBase = 0x1F0;
-					*ControlPortBase = 0x3F6;
-					*InterruptVector = 14;
-					ret = STATUS_SUCCESS;
-				}
-				break;
-			case 1:
-				if (IoGetConfigurationInformation()->AtDiskSecondaryAddressClaimed)
-					ret = STATUS_INSUFFICIENT_RESOURCES;
-				else
-				{
-					*CommandPortBase = 0x170;
-					*ControlPortBase = 0x376;
-					*InterruptVector = 15;
-					ret = STATUS_SUCCESS;
-				}
-				break;
-		}
-	}
-
-	return ret;
+    PAGED_CODE();
+
+    if (PdoExtension->Common.PageFiles ||
+        PdoExtension->Common.HibernateFiles ||
+        PdoExtension->Common.DumpFiles)
+    {
+        return STATUS_DEVICE_BUSY;
+    }
+
+    return STATUS_SUCCESS;
 }
 
-static NTSTATUS
+static
+CODE_SEG("PAGE")
+NTSTATUS
+PciIdeXPdoQueryTargetDeviceRelations(
+    _In_ PPDO_DEVICE_EXTENSION PdoExtension,
+    _In_ PIRP Irp)
+{
+    PDEVICE_RELATIONS DeviceRelations;
+
+    PAGED_CODE();
+
+    DeviceRelations = ExAllocatePoolWithTag(PagedPool,
+                                            sizeof(DEVICE_RELATIONS),
+                                            TAG_PCIIDEX);
+    if (!DeviceRelations)
+        return STATUS_INSUFFICIENT_RESOURCES;
+
+    DeviceRelations->Count = 1;
+    DeviceRelations->Objects[0] = PdoExtension->Common.Self;
+    ObReferenceObject(PdoExtension->Common.Self);
+
+    Irp->IoStatus.Information = (ULONG_PTR)DeviceRelations;
+    return STATUS_SUCCESS;
+}
+
+static IO_COMPLETION_ROUTINE PciIdeXOnRepeaterCompletion;
+
+static
+NTSTATUS
+NTAPI
+PciIdeXOnRepeaterCompletion(
+    _In_ PDEVICE_OBJECT DeviceObject,
+    _In_ PIRP Irp,
+    _In_reads_opt_(_Inexpressible_("varies")) PVOID Context)
+{
+    UNREFERENCED_PARAMETER(DeviceObject);
+
+    if (Irp->PendingReturned)
+        KeSetEvent((PKEVENT)Context, IO_NO_INCREMENT, FALSE);
+
+    return STATUS_MORE_PROCESSING_REQUIRED;
+}
+
+static
+CODE_SEG("PAGE")
+NTSTATUS
+PciIdeXPdoRepeatRequest(
+    _In_ PPDO_DEVICE_EXTENSION PdoExtension,
+    _In_ PIRP Irp,
+    _In_opt_ PDEVICE_CAPABILITIES DeviceCapabilities)
+{
+    PDEVICE_OBJECT Fdo = PdoExtension->ParentController->Common.Self;
+    PDEVICE_OBJECT TopDeviceObject = IoGetAttachedDeviceReference(Fdo);
```

Comment:
Nit: I'd prefer this to be away from the initializers and down with the "real" code since it has a notable side effect (the reference) that shouldn't be hidden away.